sys/include/usb/usbus.h

```diff
@@ -0,0 +1,515 @@
+    struct usbus_interface_alt *next; /**< Next alternative setting */
+    usbus_hdr_gen_t *hdr_gen;         /**< Optional additional header
+                                           generator */
+    usbus_endpoint_t *ep;             /**< List of associated endpoints for
+                                           this alternative setting */
+} usbus_interface_alt_t;
+
+/**
+ * @brief USBUS interface
+ */
+typedef struct usbus_interface {
+    struct usbus_interface *next;     /**< Next interface (set by USBUS during
+                                           registration)                     */
+    struct usbus_interface_alt *alts; /**< List of alt settings              */
+    usbus_handler_t* handler;         /**< Handlers for this interface       */
+    usbus_hdr_gen_t *hdr_gen;         /**< Optional additional header
+                                           generators                        */
+    usbus_endpoint_t *ep;             /**< Linked list of endpoints belonging
+                                           to this interface                 */
+    usbus_string_t *descr;            /**< Descriptor string                 */
+    uint16_t idx;                     /**< Interface index, (set by USBUS
+                                           during registration               */
+    uint8_t class;                    /**< USB interface class               */
+    uint8_t subclass;                 /**< USB interface subclass            */
+    uint8_t protocol;                 /**< USB interface protocol            */
+} usbus_interface_t;
+
+/**
+ * @brief USBUS event handler function pointers
+ */
+typedef struct usbus_handler_driver{
+
+    /**
+     * @brief Initialize the event handler
+     *
+     * This function is called in the USBUS thread context to initialize the
+     * event handler
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     */
+    void (*init)(usbus_t *usbus, struct usbus_handler *handler);
+
+    /**
+     * @brief event handler function
+     *
+     * This function is passed USBUS events
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     * @param event     @ref USBUS_MSG_TYPES "event" to handle
+     * @param arg       Additional argument
+     */
+    int (*event_handler)(usbus_t *usbus, struct usbus_handler *handler, uint16_t event, void *arg);
+} usbus_handler_driver_t;
+
+/**
+ * @ USBUS handler struct
+ *
+ * Inherit from this struct for custom USB functionality
+ */
+struct usbus_handler {
+    struct usbus_handler *next;           /**< List of handlers (to be used by
+                                               @ref usbus_t) */
+    const usbus_handler_driver_t *driver; /**< driver for this handler */
+    usbus_interface_t *iface;             /**< Interface this handler belongs
+                                               to */
+    uint32_t flags;                       /**< Report flags */
```

Comment:
Maybe learn from [previous discussion](https://github.com/RIOT-OS/RIOT/pull/10214#discussion_r260477918) and define a `usbus_handler_flags_t` for this